### ISSUE TITLE
Always pass dataloader_idx to step functions.

### ISSUE THREAD

**Issue by jopo666** [PyTorchLightning/pytorch-lightning#10315]
## Proposed refactoring or deprecation

Pass `dataloader_idx` to `[train,test,val,inference]_step` regardless of the number dataloaders.

### Motivation

I often use the same lightning modules in different projects, ie. I have one `Classifier` task that I use when I need to train classifiers. Now `Trainer` passes either 3 or 4 arguments to `step` functions depending on whether I have one or multiple loaders.

```python
# Fails with multiple dataloaders.
def training_step(self, batch, batch_idx):
    return  self(batch).sum()

# Fails if with only one DataLoader.
def training_step(self, batch, batch_idx, dataloader_idx):
    return  self(batch).sum()
```

A quick fix that could be used (ie. I'm using) is to define the training step as below but it's not pretty... :D

```python
def training_step(self, batch, batch_idx, *args):
    dataloader_idx = 0 if len(args) == 0 else args[0]
```

I think passing the dataloader_idx anyways to step functions would simplify `LightningModule` as the module itself wouldn't have to be changed depending on the number of dataloaders.

### Pitch

Always pass `dataloader_idx` to `step` functions.

**Comment by jopo666** [comment]
Here's the bug report model demonstrating the issue.

```python
import os

import torch
from torch.utils.data import DataLoader, Dataset

from pytorch_lightning import LightningModule, Trainer


class RandomDataset(Dataset):
    def __init__(self, size, length):
        self.len = length
        self.data = torch.randn(length, size)

    def __getitem__(self, index):
        return self.data[index]

    def __len__(self):
        return self.len


class BoringModel(LightningModule):
    def __init__(self):
        super().__init__()
        self.layer = torch.nn.Linear(32, 2)

    def forward(self, x):
        return self.layer(x)

    def training_step(self, batch, batch_idx):
        loss = self(batch).sum()
        self.log("train_loss", loss)
        return {"loss": loss}

    def validation_step(self, batch, batch_idx, dataloader_idx):
        loss = self(batch).sum()
        self.log("valid_loss", loss)

    def test_step(self, batch, batch_idx):
        loss = self(batch).sum()
        self.log("test_loss", loss)

    def configure_optimizers(self):
        return torch.optim.SGD(self.layer.parameters(), lr=0.1)


def run():
    train_data = DataLoader(RandomDataset(32, 64), batch_size=2)
    val_data_1 = DataLoader(RandomDataset(32, 64), batch_size=2)
    val_data_2 = DataLoader(RandomDataset(32, 64), batch_size=2)

    model = BoringModel()
    trainer = Trainer(
        default_root_dir=os.getcwd(),
        limit_train_batches=1,
        limit_val_batches=1,
        num_sanity_val_steps=0,
        max_epochs=1,
    )
    # This runs okay.
    trainer.fit(model, train_dataloaders=train_data, val_dataloaders=[val_data_1, val_data_2])
    # This fails.
    trainer.fit(model, train_dataloaders=train_data, val_dataloaders=val_data_1))


if __name__ == "__main__":
    run()

```

**Comment by carmocca** [comment]
You can define the argument with a default to avoid the error.

`def validation_step(self, batch, batch_idx, dataloader_idx=0):`
or
`def validation_step(self, batch, batch_idx, dataloader_idx=None):`

**Comment by jopo666** [comment]
Ahh perfect! Thanks also for the pull request.

**Comment by flourish-727** [comment]
> You can define the argument with a default to avoid the error.
> 
> `def validation_step(self, batch, batch_idx, dataloader_idx=0):` or `def validation_step(self, batch, batch_idx, dataloader_idx=None):`

Hello Carmocca, 

I am sorry for replying on this old post. 
But as a rookie with Lightning, I don't really understand the purpose of "batch_idx" as an argument of the methods like "training_step" or validation_step.
The method does not really use the argument (see this example from official):

def training_step(self, batch, batch_idx):
x, y, z = batch
out = self.encoder(x)
loss = self.loss(out, x)
return loss

See? "batch_idx" goes nowhere!!!
Could you please give me some hints to understand it?
Thanks!

**Comment by carmocca** [comment]
@flourish-727 You don't have to use it. It is most often not used. You can also remove it from the signature and it should still work with the latest Lightning version

**Comment by flourish-727** [comment]
> @flourish-727 You don't have to use it. It is most often not used. You can also remove it from the signature and it should still work with the latest Lightning version

Thanks a lot for the response!
But why the parameter "dataloader_idx" is a required parameter when multiple Dataloader instances are used. 
For example, "dataloader_idx" is not used in the body in the example below, but we must put it in the signature. 
Otherwise, it is going to give an error with multiple Dataloader instances.

def validation_step(self, batch, batch_idx, dataloader_idx=0):
        loss = self(batch).sum()
        self.log("valid_loss", loss)

I would assume trainer.fit() method use it somewhere somehow under the hood. 
But I just cannot figure out how it makes that happen if I don't use "dataloader_idx" in the body of the method at all. 
I hope you understand what I am trying to ask. 
I am not using "dataloader_idx" in the body but it is required in the method signature for multiple Dataloader instances, which is weird to me. 
Thanks for any help!
